File: packages/common/faultline_common/log.py
"""Unified JSON-lines logging used by every Faultline service.

One line per event on stdout so Modal captures it (`modal app logs <app> -e local`).
Shape (stable, keep in sync with apps/web/src/lib/log.ts):

    {"ts": ISO8601Z, "svc": "harness|sandbox-env|web|script", "lvl": "debug|info|warn|error",
     "run_id"?: str, "episode_id"?: str, "step"?: int, "ev": "dotted.event.name", "msg": str, ...extras}

Set LOG_LEVEL=debug to include argument/output dumps (callers truncate them first).
"""

from __future__ import annotations

import contextvars
import json
import os
import sys
import time
from datetime import datetime, timezone
from typing import Any

_LEVELS = {"debug": 10, "info": 20, "warn": 30, "error": 40}

# Correlation ids that services set once per request/episode so every line carries them.
ctx_run_id: contextvars.ContextVar[str | None] = contextvars.ContextVar("run_id", default=None)
ctx_episode_id: contextvars.ContextVar[str | None] = contextvars.ContextVar("episode_id", default=None)
ctx_step: contextvars.ContextVar[int | None] = contextvars.ContextVar("step", default=None)
ctx_request_id: contextvars.ContextVar[str | None] = contextvars.ContextVar("request_id", default=None)


def _now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _min_level() -> int:
    return _LEVELS.get(os.environ.get("LOG_LEVEL", "info").lower(), 20)


def truncate(s: Any, limit: int = 2000) -> Any:
    """Bound a string for log/event payloads, marking truncation explicitly."""
    if not isinstance(s, str) or len(s) <= limit:
        return s
    return s[:limit] + f"…[truncated {len(s) - limit} chars]"


def log_event(svc: str, ev: str, msg: str = "", lvl: str = "info", **extras: Any) -> dict[str, Any]:
    """Emit one JSON log line. Returns the record (handy for mirroring into event streams)."""
    if _LEVELS.get(lvl, 20) < _min_level():
        return {}
    rec: dict[str, Any] = {"ts": _now(), "svc": svc, "lvl": lvl}
    for key, var in (("run_id", ctx_run_id), ("episode_id", ctx_episode_id), ("step", ctx_step), ("request_id", ctx_request_id)):
        val = var.get()
        if val is not None:
            rec[key] = val
    rec["ev"] = ev
    rec["msg"] = msg
    for k, v in extras.items():
        if v is not None:
            rec[k] = v
    sys.stdout.write(json.dumps(rec, default=str, ensure_ascii=False) + "\n")
    sys.stdout.flush()
    return rec


class Logger:
    """Tiny bound logger: `log = get_logger("sandbox-env"); log.info("tool.result", "ok", tool="read_file")`."""

    def __init__(self, svc: str):
        self.svc = svc

    def debug(self, ev: str, msg: str = "", **kw: Any) -> dict[str, Any]:
        return log_event(self.svc, ev, msg, "debug", **kw)

    def info(self, ev: str, msg: str = "", **kw: Any) -> dict[str, Any]:
        return log_event(self.svc, ev, msg, "info", **kw)

    def warn(self, ev: str, msg: str = "", **kw: Any) -> dict[str, Any]:
        return log_event(self.svc, ev, msg, "warn", **kw)

    def error(self, ev: str, msg: str = "", **kw: Any) -> dict[str, Any]:
        return log_event(self.svc, ev, msg, "error", **kw)

    class _Timer:
        def __init__(self, outer: "Logger", ev: str, **kw: Any):
            self.outer, self.ev, self.kw = outer, ev, kw
            self.t0 = time.perf_counter()

        def __enter__(self) -> "Logger._Timer":
            return self

        def __exit__(self, *exc: Any) -> None:
            dur = int((time.perf_counter() - self.t0) * 1000)
            if exc[0] is not None:
                self.outer.error(self.ev, f"failed: {exc[1]}", dur_ms=dur, **self.kw)
            else:
                self.outer.info(self.ev, "ok", dur_ms=dur, **self.kw)

    def timed(self, ev: str, **kw: Any) -> "Logger._Timer":
        return Logger._Timer(self, ev, **kw)


def get_logger(svc: str) -> Logger:
    return Logger(svc)

File: apps/web/src/lib/log.ts
/**
 * Unified JSON-lines logging for the browser — same shape as
 * packages/common/faultline_common/log.py, with svc:"web".
 *
 *   {"ts":ISO8601Z,"svc":"web","lvl":"info","run_id"?:str,"episode_id"?:str,"step"?:int,
 *    "ev":"dotted.event.name","msg":str, ...extras}
 *
 * Lines go to the console as a single JSON string (so a reviewer can copy/paste them next to the
 * platform log capture) AND into a bounded in-memory ring buffer that the Logs tab renders.
 */

import type { LogLevel, LogLine } from './types'

const LEVELS: Record<LogLevel, number> = { debug: 10, info: 20, warn: 30, error: 40 }

const BUFFER_LIMIT = 2000

const minLevel: number = LEVELS.info
let buffer: LogLine[] = []
let seq = 0
const subscribers = new Set<(lines: LogLine[]) => void>()

export function nowIso(): string {
  return new Date().toISOString().replace(/(\.\d{3})\d*Z$/, '$1Z')
}

function notify(): void {
  const snapshot = buffer
  for (const fn of subscribers) fn(snapshot)
}

/** Append an already-formed unified log line (e.g. a `log` event mirrored from the harness). */
export function ingest(line: LogLine): void {
  buffer = [...buffer, { ...line, _seq: seq++ }].slice(-BUFFER_LIMIT)
  notify()
}

export function subscribeLogs(fn: (lines: LogLine[]) => void): () => void {
  subscribers.add(fn)
  fn(buffer)
  return () => {
    subscribers.delete(fn)
  }
}

const CONSOLE: Record<LogLevel, (msg: string) => void> = {
  debug: (m) => console.debug(m),
  info: (m) => console.info(m),
  warn: (m) => console.warn(m),
  error: (m) => console.error(m),
}

export function logEvent(
  svc: string,
  ev: string,
  msg = '',
  lvl: LogLevel = 'info',
  extras: Record<string, unknown> = {},
): LogLine | null {
  if (LEVELS[lvl] < minLevel) return null
  const rec: LogLine = { ts: nowIso(), svc, lvl, ev, msg }
  for (const [k, v] of Object.entries(extras)) {
    if (v !== undefined && v !== null) rec[k] = v
  }
  try {
    CONSOLE[lvl](JSON.stringify(rec))
  } catch {
    // JSON.stringify can throw on cyclic extras; never let logging break the app.
    CONSOLE[lvl](`{"ts":"${rec.ts}","svc":"${svc}","lvl":"${lvl}","ev":"${ev}","msg":"${msg}"}`)
  }
  ingest(rec)
  return rec
}

export class Logger {
  constructor(private readonly svc: string) {}

  debug = (ev: string, msg = '', extras: Record<string, unknown> = {}) =>
    logEvent(this.svc, ev, msg, 'debug', extras)

  info = (ev: string, msg = '', extras: Record<string, unknown> = {}) =>
    logEvent(this.svc, ev, msg, 'info', extras)

  warn = (ev: string, msg = '', extras: Record<string, unknown> = {}) =>
    logEvent(this.svc, ev, msg, 'warn', extras)

  error = (ev: string, msg = '', extras: Record<string, unknown> = {}) =>
    logEvent(this.svc, ev, msg, 'error', extras)
}

export function getLogger(svc: string): Logger {
  return new Logger(svc)
}
